src/System.Management.Automation/engine/hostifaces/Command.cs

```diff
@@ -497,13 +509,27 @@ CommandOrigin origin
                     commandProcessorBase = new CommandProcessor(functionInfo, executionContext,
                                                                 _useLocalScope ?? false, fromScriptFile: false, sessionState: executionContext.EngineSessionState);
                 }
+                else if (ScriptPath != null)
+                {
+                    var scriptName = Path.GetFileName(ScriptPath);
+                    var scriptInfo = new ExternalScriptInfo(scriptName, ScriptPath, executionContext);
+                    commandProcessorBase = new DlrScriptCommandProcessor(
+                        scriptInfo,
+                        executionContext,
+                        _useLocalScope ?? false,
+                        executionContext.EngineSessionState);
+
+                    commandProcessorBase.Command.MyInvocation.InvocationName = ScriptPath;
+                }
                 else
                 {
-                    commandProcessorBase = new DlrScriptCommandProcessor(scriptBlock,
-                                                                         executionContext, _useLocalScope ?? false,
-                                                                         origin,
-                                                                         executionContext.EngineSessionState,
-                                                                         DollarUnderbar);
+                    commandProcessorBase = new DlrScriptCommandProcessor(
+                        scriptBlock,
+                        executionContext,
+                        _useLocalScope ?? false,
+                        origin,
```

Comment:
Why do we ignore this above if `ScriptPath != null`?

Reply:
I am not sure I follow. If we have path to the script we can use the same `DlrScriptCommandProcessor` constructor as would be used if the file had .ps1 extension. What is being ignored in your opinion?

Reply:
The constructor that gets passed `ExternalScriptInfo` as the first argument is necessary for the `InvocationInfo` (`$MyInvocation` special var) to be created properly.

Reply:
My ask is about `origin` argument.

Reply:
I guess the answer to this question is the same as answer to question why `DlrScriptCommandProcessor` constructor for external script does not accept `origin` parameter? In my opinion it's because it is not need at this stage where we already know the command is external script.

Note that the `if` above for cases when `scriptBlock.UsesCmdletBinding` is `true` also does not pass the `origin` to the `CommandProcessor` constructor so it's the same case in that regard. Isn't it because `CommandOrigin` is needed only in special cases and during command searching/discovery?

I have seen PowerShell's source code for the first time a few days ago so I definitely do not feel competent enough to answer such questions but I guess that when it gets looked at by somebody who knows the inner workings really well he/she will be immediately able to tell whether it's secure to do it this way. In my opinion it should be because the code path that gets executed when the script file has the `.ps1` extension does exactly the same (in addition to command discovery which is redundant in our case).

Reply:
Thanks! I did not look in depth but I'd still expect that `origin` should be here.

Reply:
Just looked into this. I'll say that the constructor code here is old and frankly hard to follow (could use some style improvements).

But when the `CommandInfo` is provided, the origin is taken to be `Internal`:

https://github.com/PowerShell/PowerShell/blob/1223518a437ccc06c49357358150919db9b91eab/src/System.Management.Automation/engine/ScriptCommandProcessor.cs#L38